### ISSUE TITLE
Incorrect evaluation of 'Directory.Packages.props'

### ISSUE THREAD

**Issue by shlomiassaf** [leonardochaia/dotnet-affected#54]
The logic for processing `Directory.Packages.props` in a central package versioning report does not follow the spec:

https://github.com/leonardochaia/dotnet-affected/blob/9893044cf8a026f2f7c700bfa8a7128e05126790/src/dotnet-affected/Domain/AffectedExecutor.cs#L102-L103

It assumes a single `Directory.Packages.props` can exist in a repo.

If there are multiple, and both have changed in the same run it will throw.

From Microsoft [docs](https://docs.microsoft.com/en-us/dotnet/core/compatibility/sdk/5.0/directory-packages-props-imported-by-default):
>NuGet's .props files automatically import a file named Directory.Packages.props if it's found in the project folder or any of its ancestors.

 I.E the **first** `Directory.Packages.props` file found in the project's directory or any other folder up to the root (solution directory) is used. If a file is found and there are additional files in ancestor folders they are ignored by the build.

Usually, one will build nested `Directory.Packages.props` files like this:

```xml
<Project>

    <!-- First import top level package props... -->

    <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory).., Directory.Packages.props))\Directory.Packages.props" />

    <!-- Override/Add with project local versioning -->

    <ItemGroup>
        <PackageVersion Include="System.Text.Json" Version="6.0.4" />
    </ItemGroup>
</Project>
```

Since MS Build `Project` instance is used to load the package props:
https://github.com/leonardochaia/dotnet-affected/blob/9893044cf8a026f2f7c700bfa8a7128e05126790/src/dotnet-affected/Infrastructure/NugetHelper.cs#L26

I can only assume the only file we need to load is the one closest to the project folder and ignore (but not fail) others.

For the a 1st phase fix this will suffice.

It is however, not 100% accurate, but a bit more complex to perfect it... I'll elaborate in a new comment below

**Comment by shlomiassaf** [comment]
So, for a 2nd phase...

This is a bit tricky since we first need to see if there is any `Directory.Packages.props` file that has changed.  
If so, for each project we need to see evaluate the first `Directory.Packages.props` and load it, even if it did not change but a parent of it has changed.

It means that 

1) A `Directory.Packages.props` that changed inside project A should not trigger package change in project B since project B does not "know" about that file. It only knows about the solution's `Directory.Packages.props` or its own if it has one.

2) If project A has it's own `Directory.Packages.props`, changing a package in the root `Directory.Packages.props` does not mean project A has changed.

If `Directory.Packages.props` in A reference `Directory.Packages.props` in the root, the version that changed in the root
might be overridden in `Directory.Packages.props` of project A so the final resolution means no change in A but might be changes in other projects.

Again, this is complex, I suggest fixing phase 1 first, so at least it's accurate in a level that does not miss changes, it might only do redundant work however in rare edge cases so it's something you can live with

**Comment by leonardochaia** [comment]
Hi @shlomiassaf ! thanks for bringing this up. I'm sure there are other things that are not up to spec as well.

Finding the `Directory.Packages.props` (DPP for short?) that belongs to each project when using more than one DPP is not hard. Each `Project` has a `Property` that tells you the path to the DPP for that project. The prediction library we use will also output the correct DPP as input for each project.

So we can fix the finding of the correct DPP, that should get us in the right way. It gets tricky when packages are declared in more than one file.
I'm not sure if the 2nd phase you mention, the prediction library would output both files as input. That would be great if that's the case.

**Comment by shlomiassaf** [comment]
@leonardochaia I'm not sure it's thats easy.

I'll start with some background and explanation, below is my proposed solution...

Basically, at the moment the library support's DPP indirectly, i.e. it explicitly search for DPP file and load it as a `Project` instance.

It is not "native" so it create's challenges which is what I wrote about "following the spec".

Basically, the most efficient path is to only load `csproj` files (or any lang proj file) and let it do the magic. When we load `csproj` it will load everything for us.  

As it stand, for DPP to work the library does some specific work.  
- It will search for the DPP file
- It checks for `ManagePackageVersionsCentrally` and implement the logic for it ()
- It will compare it to the package in the project files manually

There are a lot of issues with the above because you manually implement the resolution logic for DPP, but with missing features (like imports...)  
One key issue is that for truly implementing it you must always load the DPP closest to the `csproj`.  

Remember DPP might not import from a parent, some might and you also need to detect which through git changes (multiple DPP, parent DPP over project etc...)

All can be avoided just by importing the `csproj` directly as a `Project` instance which **WILL** load the DPP, Build.props, targets and what not... full, real resolutions.

The problem with loading `csproj` files is of course the file system, and this is where the solution is...

----

The idea is to allow `Project` instance to load while having it resolve the files from where we want.  
Right now the library does that only for the first file loaded into `Project` and other files are just loaded relative to the root 
the command ran from. (PWD)

When using the `working directory` that's quite easy, however not supported currently, but a very quick fix.

The complexity comes when we need to load it from a commit (HEAD or any other commit).  

There is a solution, which is supported in build version 16.11 and above, i.e `net5.0`, `net6.0` and onwards.   
So the old logic will be left in 3.1 but moving forward we can use the new solution.

From 16.11, build accepts an `IFileSystem` implementation, used by the build system when it want to load files.  
From here, you can quickly virtualize a file system to a specific state represented by a commit.   

I have a fork i'm working on with this solution, I'll post a PR soon.

I'm not sure I'll do the whole `csproj` implementation in the first PR as it's just to much.  
First having the file system working and being able to resolve the proper files when loading a `Project`

I'll post it soon

**Comment by shlomiassaf** [comment]
https://github.com/shlomiassaf/dotnet-affected/tree/support-import-element

It's not done

**Comment by shlomiassaf** [comment]
So, loading the `csproj` as a whole will not work ATM

https://github.com/dotnet/msbuild/issues/7956

We can workaround by preloading it to the cache but that will require some clever hacks and knowing when we need to pre-load from our internal directory.

Let's wait for their thoughts.
It's not blocking the PR though, it just means we will still analyse DPP files manually.

**Comment by leonardochaia** [comment]
Hey Shlomi! I am catching up with your comments & PR.

> All can be avoided just by importing the csproj directly as a Project instance WILL load the DPP, Build.props, targets and what not... full, real resolutions.

This is great!! I ofc agree that we should leverage most we can to MSBuild and avoid having custom re-implementation of the spec.

There's something I still don't get tho. Once the `Project` is loaded, including the props/targets files, is there a way using MSBuild to "query which packages a project is referencing" (including DPP files, etc)? 

Having a virtual FS for the git references we are comparing is awesome!! Taking a look at the PR now.

**Comment by shlomiassaf** [comment]
> There's something I still don't get tho. Once the Project is loaded, including the props/targets files, is there a way using MSBuild to "query which packages a project is referencing" (including DPP files, etc)?

Once a `Project` is loaded you have all the data you needs, properies, metadata etc... all evaluated and unevaluated.

You have it all aggregated to the project instance, from all files, after loading them internally.

I already have a branch (in my fork) with a good workaround for the MS issue I opened, where I just use a single `csproj` file to evaluate all the things I need.

At the moment I still work directly with `PackageReference` and `PackageVersion` and I still need to check the propeties related to DPP, like if it's disabled, etc... but in one single file.

I've also added there support for `OverrideVersion` and the check for disabling `OverrideVersion`.

I did not find the microsoft code/library that based on a project instance returns the list of resolved packages.
It's out there, it might be something for the future but either way, that will resolve the packages **after** evaluation, i.e. conditions will apply so it might not be the desired solution when we want to ignore confitions.
